pan3d/explorers/analytics.py

```diff
@@ -0,0 +1,576 @@
+            local_rendering (str): If provided (wasm, vtkjs) local rendering will be used
+
+        CLI options:
+            - `--import-state`: Pass a string with this argument to specify a startup configuration.
+                              This value must be a local path to a JSON file which adheres to the
+                              schema specified in the [Configuration Files documentation](../api/configuration.md).
+                              A dataset specified in this configuration will override any value passed to `--xarray-*`
+            - `--xarray-file`: Provide path to xarray file
+            - `--xarray-url`: Provide URL to xarray dataset
+            - `--wasm`: Use WASM for local rendering
+            - `--vtkjs`: Use vtk.js for local rendering
+        """
+        self.xarray = xarray
+        self.server = get_server(server, client_type="vue3")
+        if self.server.hot_reload:
+            self.ctrl.on_server_reload.add(self._build_ui)
+
+        # cli
+        self.server.cli.add_argument(
+            "--import-state",
+            help="Pass a string with this argument to specify a startup configuration. This value must be a local path to a JSON file which adheres to the schema specified in the [Configuration Files documentation](../api/configuration.md). A dataset specified in this configuration will override any value passed to `--xarray-*`",
+        )
+        self.server.cli.add_argument(
+            "--xarray-file",
+            help="Provide path to xarray file",
+        )
+        self.server.cli.add_argument(
+            "--xarray-url",
+            help="Provide URL to xarray dataset",
+        )
+
+        # Local rendering setup
+        self.server.cli.add_argument(
+            "--wasm",
+            help="Use WASM for local rendering",
+            action="store_true",
+        )
+        self.server.cli.add_argument(
+            "--vtkjs",
+            help="Use vtk.js for local rendering",
+            action="store_true",
+        )
+        args, _ = self.server.cli.parse_known_args()
+        self.local_rendering = local_rendering
+        if args.wasm:
+            self.local_rendering = "wasm"
+        if args.vtkjs:
+            self.local_rendering = "vtkjs"
+
+        # If no GPU, use local rendering
+        if self.local_rendering is None and not has_gpu():
+            self.local_rendering = "wasm"
+
+        # Process CLI
+        self.ctrl.on_server_ready.add(self._process_cli)
+
+        self.state.nan_colors = [
+            [0, 0, 0, 1],
+            [0.99, 0.99, 0.99, 1],
+            [0.6, 0.6, 0.6, 1],
+            [1, 0, 0, 1],
+            [0, 1, 0, 1],
+            [0, 0, 1, 1],
+            [0.9, 0.9, 0.9, 0],
+        ]
+        self.state.nan_color = 2
+
+        self.ui = None
+        self._setup_vtk()
+        self._build_ui()
+        print("forcing update figure", self.state.color_by)
+        self.plotting.on_change_active_plot()
```

Comment:
This should automatically be called right before on_server_ready as there is a @state.change attached to it. You should not need to make that call here.

Reply:
We could create a pan3d abstract class for an "app". That way we can get the boiler plate code done for all.